src/Tools/dotnet-monitor/Egress/EgressProviderSource.cs

```diff
@@ -0,0 +1,103 @@
+using Microsoft.Diagnostics.Tools.Monitor.Extensibility;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Primitives;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.Egress
+{
+    internal sealed class EgressProviderSource : IDisposable
+    {
+        private readonly Lazy<IDisposable> _changeRegistrationLazy;
+        private readonly IEgressConfigurationProvider _configurationProvider;
+        private readonly ExtensionDiscoverer _extensionDiscoverer;
+        private readonly ILogger _logger;
+        private readonly IDictionary<string, string> _providerNameToTypeMap;
+
+        public EgressProviderSource(
+            IEgressConfigurationProvider configurationProvider,
+            ExtensionDiscoverer extensionDiscoverer,
+            ILogger<EgressProviderSource> logger)
+        {
+            _changeRegistrationLazy = new Lazy<IDisposable>(CreateChangeRegistration, LazyThreadSafetyMode.ExecutionAndPublication);
+            _configurationProvider = configurationProvider;
+            _extensionDiscoverer = extensionDiscoverer;
+            _logger = logger;
+            _providerNameToTypeMap = new ConcurrentDictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+        }
+
+        public void Dispose()
+        {
+            if (_changeRegistrationLazy.IsValueCreated)
+            {
+                _changeRegistrationLazy.Value.Dispose();
+            }
+        }
+
+        public void Initialize()
+        {
+            _ = _changeRegistrationLazy.Value;
+        }
+
+        public IEgressExtension GetEgressProvider(string name)
+        {
+            if (!_providerNameToTypeMap.TryGetValue(name, out string providerType))
+            {
+                throw new EgressException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressProviderDoesNotExist, name));
+            }
+
+            return _extensionDiscoverer.FindExtension<IEgressExtension>(providerType);
+        }
+
+        private IDisposable CreateChangeRegistration()
+        {
+            IDisposable changeRegistration = ChangeToken.OnChange(
+                _configurationProvider.GetReloadToken,
+                Reload);
+
+            Reload();
+
+            return changeRegistration;
+        }
+
+        private void Reload()
+        {
+            _providerNameToTypeMap.Clear();
+
+            foreach (string providerType in _configurationProvider.ProviderTypes)
+            {
+                try
+                {
+                    IEgressExtension _ = _extensionDiscoverer.FindExtension<IEgressExtension>(providerType);
+                }
+                catch (ExtensionException)
+                {
+                    _logger.EgressProviderTypeNotExist(providerType);
```

Comment:
I effectively get two of the same error message right now:
`08:00:43 fail: Microsoft.Diagnostics.Tools.Monitor.Extensibility.ExtensionDiscoverer[86]
      Failed to locate extension 'AzureBlobStorage_dotnet_tools'.
08:00:43 warn: Microsoft.Diagnostics.Tools.Monitor.Egress.EgressProviderSource[96]
      Egress provider type 'AzureBlobStorage_dotnet_tools' does not exist.`

I suppose there isn't any harm in this - it's just redundant for these to both print out without providing any unique information.

Reply:
Actually, I'd like to defer this to a different PR if actually necessary. I think the logging for extension discovery should be simplified or replaced with exceptions but don't want to complicate this PR further.